File: crash_consistency/crash_cons_pmem.cpp
#include <iostream>
#include <fstream>
#include <algorithm>
#include <math.h>
#include <pthread.h>
#include <libpmemobj/atomic_base.h>
#include <libpmemobj++/make_persistent.hpp>
#include <libpmemobj++/make_persistent_atomic.hpp>
#include <libpmemobj++/transaction.hpp>
#include <libpmemobj++/utils.hpp>
#include <stdint.h>
#include "../pvector.hpp"
#define CREATE_MODE_RW (S_IWUSR | S_IRUSR)


using namespace std;
using namespace pmem::obj;

const char *LAYOUT = "fkrcplus";
const int64_t iterations = 50000;
const int64_t MAX_CAP = 50000;
const int64_t default_cap = 2500;
const uint64_t PMEM_FILE_SZ = (uint64_t)3*1024*1024*1024;

struct root {
    persistent_ptr<PVector<int64_t>> arr;
    int64_t iter;
};

pmem::obj::pool<root> pop;

bool file_exists(const std::string &path) {
   std::ifstream f(path.c_str());
   return f.good();
}

int main(int argc, char **argv)
{

    std::string path = "pmemtest";

    if (file_exists(path)) {
        std::cout << "opening existing pmem file\n";
        pop = pmem::obj::pool<root>::open(path, LAYOUT);
        if (pop.root()->iter == iterations) {
            transaction::run(pop, [&] {
                pop.root()->iter = 0;
                pop.root()->arr->reset(default_cap);
            });
        }
    } else {
        std::cout << "opening new pmem file\n";
        pop = pmem::obj::pool<root>::create(path, LAYOUT, PMEM_FILE_SZ, CREATE_MODE_RW);
        transaction::run(pop, [&] {
            pop.root()->iter = 0;
            pop.root()->arr = make_persistent<PVector<int64_t>>(default_cap);
        });
    }

    auto root = pop.root();

    while (root->iter < iterations) {
        transaction::run(pop, [&root] {
            int64_t sum = 0;
            for (int64_t i = 0; i < root->arr->cap(); i++) {
                int64_t val = (int64_t)(ceil(sqrt(root->arr->get(i)) + (((i + 1) * root->arr->cap()) % MAX_CAP)));
                root->arr->set(i, val);
                sum = sum + val;
            }
            int64_t newcap = (sum % MAX_CAP) + 1;
            if (newcap <= 0) {
                newcap = default_cap;
            }
            root->arr->resize(newcap);
            root->iter += 1;
            if (root->iter == iterations) cout << sum << endl;
        });
    }
    pop.close();

    return 0;
}

File: crash_consistency/crash_cons_pmem.go
package main

import (
   "flag"
   "math"
   "github.com/vmware/go-pmem-transaction/pmem"
   "github.com/vmware/go-pmem-transaction/transaction"
)
var pmemFile = flag.String("file", "pmemtest", "pmemtest file name")

type PVector struct {
	capacity int32
	arr []int32
}

func Init(cap int32) *PVector {
    var pv *PVector
    pv = (*PVector)(pmem.Get("pvec", pv))

    var c int = 2500

    if pv == nil {
        pv = (*PVector)(pmem.New("pvec", pv))
        pv.capacity = cap
        pv.arr = pmem.Make("pvec_slice", pv.arr, c).([]int32)
        return pv
    }

    return pv
} 

func (pv *PVector) cap() int32 {
    return pv.capacity
}

func (pv *PVector) get(i int32) int32 {
    return pv.arr[i]
}

func (pv *PVector) set(i int32, v int32) {
    pv.arr[i] = v
}

func (pv *PVector) resize(sz int32) {
    var temp []int32 = make([]int32, pv.capacity)
    for i := int32(0); i < pv.capacity; i++ {
        temp[i] = pv.arr[i]
    }

    pmem.Delete("pvec_slice")
    pv.arr = pmem.Make("pvec_slice", pv.arr, sz).([]int32)

    var ec int32
    if sz > pv.capacity {
        ec = pv.capacity
    } else {
        ec = sz
    }

    for i := int32(0); i < ec; i++ {
        pv.arr[i] = temp[i]
    }
    pv.capacity = sz
}

func (pv *PVector) reset(sz int32) {
    pmem.Delete("pvec_slice")
    pv.arr = pmem.Make("pvec_slice", pv.arr, sz).([]int32)
}


var iterations int32 = 50000
var MAX_CAP int32 = 50000
var default_cap int32 = 2500

var pv *PVector
var iter *int32

func main() {
    flag.Parse()
    firstInit := pmem.Init(*pmemFile)
	if firstInit {
        println("new")
        pv = Init(default_cap)
        iter = (*int32)(pmem.New("iter", iter))
        *iter = 0
	} else {
        pv = (*PVector)(pmem.Get("pvec", pv))
        iter = (*int32)(pmem.Get("iter", iter))

        if pv == nil {
            println("new")
            pv = Init(default_cap)
        }
        if iter == nil {
            iter = (*int32)(pmem.New("iter", iter))
            *iter = 0
        }
        if *iter == 50000 {
            *iter = 0
            pv.reset(default_cap)
        }
    }
    
	for ; *iter < iterations; {
        txn("undo") {
            var sum int32 = 0
            for i := int32(0); i < pv.cap(); i++ {
                var val int32
                val = int32(math.Ceil(float64(math.Sqrt(float64(pv.get(i))) + float64(((i + 1) * pv.cap()) % MAX_CAP))))
                pv.set(i, val)
                sum += val
            }
            var newcap int32 = sum % MAX_CAP + 1
            if (newcap <= 0) {
                newcap = default_cap
            }
            pv.resize(newcap)
            (*iter)++
        }
	}
}